llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -211,65 +382,88 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
       Type *SCTy = CI->getType();
+      unsigned NameArgNo = 0;
+      if (IsComposite) { // structs are returned via sret arguments
+        NameArgNo = 1;
+        auto *PtrTy = cast<PointerType>(CI->getArgOperand(0)->getType());
+        SCTy = PtrTy->getElementType();
+      }
+      StringRef SymID = getStringLiteralArg(CI, NameArgNo, DelInsts);
 
       if (SetValAtRT) {
         // 2. Spec constant value will be set at run time - then add the literal
         // to a "spec const string literal ID" -> "integer ID" map, uniquing
         // the integer ID if this is new literal
         auto Ins = IDMap.insert(std::make_pair(SymID, 0));
         if (Ins.second)
-          Ins.first->second = NextID++;
-        //  3. Transform to spirv intrinsic _Z*__spirv_SpecConstant*.
-        LLVMContext &Ctx = F.getContext();
-        // Generate arguments needed by the SPIRV version of the intrinsic
-        // - integer constant ID:
-        Value *ID = ConstantInt::get(Type::getInt32Ty(Ctx), NextID - 1);
-        // - default value:
-        Value *Def = genDefaultValue(SCTy, CI);
-        // ... Now replace the call with SPIRV intrinsic version.
-        Value *Args[] = {ID, Def};
-        constexpr size_t NArgs = sizeof(Args) / sizeof(Args[0]);
-        Type *ArgTys[NArgs] = {nullptr};
-        for (unsigned int I = 0; I < NArgs; ++I)
-          ArgTys[I] = Args[I]->getType();
-        FunctionType *FT = FunctionType::get(SCTy, ArgTys, false /*isVarArg*/);
-        Module &M = *F.getParent();
-        std::string SPIRVName = mangleFuncItanium(SPIRV_GET_SPEC_CONST_VAL, FT);
-        FunctionCallee FC = M.getOrInsertFunction(SPIRVName, FT);
-        assert(FC.getCallee() && "SPIRV intrinsic creation failed");
-        CallInst *SPIRVCall =
-            CallInst::Create(FT, FC.getCallee(), Args, "", CI);
-        CI->replaceAllUsesWith(SPIRVCall);
-        // Mark the instruction with <symbolic_id, int_id> pair for later
+          Ins.first->second = NextID;
+        unsigned ID = Ins.first->second;
+
+        //  3. Transform to spirv intrinsic _Z*__spirv_SpecConstant* or
+        //  _Z*__spirv_SpecConstantComposite
+        SmallVector<unsigned, 4> GeneratedIDs;
+        auto *SPIRVCall = emitSpecConstantRecursive(ID, SCTy, GeneratedIDs, CI);
+        if (Ins.second) {
+          // emitSpecConstantRecursive might emit more than one spec constant
+          // (because of composite types) and therefore, we need to ajudst
+          // NextID according to the actual amount of emitted spec constants
+          NextID += GeneratedIDs.size();
+        }
+
+        if (IsComposite)
+          // __sycl_getCompositeSpecConstant returns through argument, so, the
+          // only thing we need to do here is to store into a memory pointed by
+          // that argument
+          new StoreInst(SPIRVCall, CI->getArgOperand(0), CI);
+        else
+          CI->replaceAllUsesWith(SPIRVCall);
```

Comment:
```suggestion
        if (IsComposite) {
          // __sycl_getCompositeSpecConstant returns through argument, so, the
          // only thing we need to do here is to store into a memory pointed by
          // that argument
          new StoreInst(SPIRVCall, CI->getArgOperand(0), CI);
        } else {
          CI->replaceAllUsesWith(SPIRVCall);
        }
```
See See https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Reply:
Thanks for the link, fixed in 4c9ce329cecc0563d51fd71aa56696fa2ff40dba